### PR TITLE
[WIP] RefChecks takes object as lazy

### LINKED ISSUE #5187: objects should override lazy values
Title says it all. Right now you cannot override a lazy val with an object. But I think you should be able to do this, or else you will be locked in forever in lazy vals.

### PR DESCRIPTION
**Pull request by som-snytt**
Fixes scala/bug#5187

### CODE REVIEW

File: src/compiler/scala/tools/nsc/typechecker/RefChecks.scala [hunk 1]
```diff
@@ -427,11 +427,11 @@ abstract class RefChecks extends Transform {
                                           "(this rule is designed to prevent accidental overrides)")
             } else if (other.isStable && !member.isStable) { // (1.4)
               overrideErrorWithMemberInfo("stable, immutable value required to override:")
-            } else if (member.isValue && member.isLazy &&
+            } else if (member.isValue && (member.isLazy || member.isModule) &&
                        other.isValue && other.hasFlag(STABLE) && !(other.isDeferred || other.isLazy)) {
               overrideErrorWithMemberInfo("concrete non-lazy value cannot be overridden:")
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/RefChecks.scala]: should we imporove this message?
```
"value must be non-lazy to override a concrete non-lazy value:"
```

File: src/compiler/scala/tools/nsc/typechecker/RefChecks.scala [hunk 2]
```diff
@@ -427,11 +427,11 @@ abstract class RefChecks extends Transform {
                                           "(this rule is designed to prevent accidental overrides)")
             } else if (other.isStable && !member.isStable) { // (1.4)
               overrideErrorWithMemberInfo("stable, immutable value required to override:")
-            } else if (member.isValue && member.isLazy &&
+            } else if (member.isValue && (member.isLazy || member.isModule) &&
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/RefChecks.scala:430]: Could we make this a deprecation warning for now, and error under `-Xsource:3`?

Comment (som-snytt) [reply src/compiler/scala/tools/nsc/typechecker/RefChecks.scala:430]: I was scared of breaking the matrix of exclusive `else ifs`.

File: src/compiler/scala/tools/nsc/typechecker/RefChecks.scala (hunk 1 above)
Comment (som-snytt) [reply src/compiler/scala/tools/nsc/typechecker/RefChecks.scala]: indeed, cut/paste the next message, add negating prefix.

File: test/files/neg/t5429.scala [hunk 3]
```diff
@@ -39,8 +39,8 @@ class C extends A {
   override object oneArg    // fail
 }
 class C0 extends A0 {
-  override object value     // !!! this succeeds, but should fail (lazy over strict)
-  override object lazyvalue // !!! this fails, but should succeed (lazy over lazy)
+  override object value     // fail
```
Comment (lrytz) [inline test/files/neg/t5429.scala:42]: it's a deprecation now, could add `// scalac: -deprecation` to the test

File: src/compiler/scala/tools/nsc/typechecker/RefChecks.scala [hunk 4]
```diff
@@ -427,11 +427,14 @@ abstract class RefChecks extends Transform {
...
                        other.isValue && other.hasFlag(STABLE) && !(other.isDeferred || other.isLazy)) {
-              overrideErrorWithMemberInfo("concrete non-lazy value cannot be overridden:")
+              if (member.isModule && !currentRun.isScala3)
+                runReporting.deprecationWarning(member.pos, member, member.owner, "overriding concrete non-lazy value with an object member will be disallowed under -Xsource:3", "2.13.11")
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/RefChecks.scala:433]: Hmm, it seems this comes too early now and prevents override checking for compatible types:

```
scala> class A { val value = 0 }
class A

scala> object O extends A { override object value }
                                            ^
       warning: overriding concrete non-lazy value with an object member will be disallowed under -Xsource:3
object O

scala> (O: A).value
java.lang.ClassCastException: class O$value$ cannot be cast to class java.lang.Integer (O$value$ is in unnamed module of loader scala.tools.nsc.interpreter.IMain$TranslatingClassLoader @475646d4; java.lang.Integer is in module java.base of loader 'bootstrap')
```

Comment (som-snytt) [reply src/compiler/scala/tools/nsc/typechecker/RefChecks.scala:433]: Sorry, I mean to revisit. I think I said I'd break the if-else.

**Pull request closed without merging**